### ISSUE TITLE
pycocotools and runner.test() scores do not match

### ISSUE THREAD

**Issue by daniellehot** [open-mmlab/mmdetection#12009]
Hi. The COCO performance as reported by *pycocotools* does not match with the reported performance from *mmdetection*. Does anybody know why? 

**Bug description** 
I execute `runner.test()` after training. This saves predictions and returns the following scores:
|Metric|IoU|Area|MaxDets|Value|
|------|---|----|-------|-----|
|Average Precision (AP)|0.50:0.95|all|100|0.120|
|Average Precision (AP)|0.50|all|100|0.281|
|Average Precision (AP)|0.75|all|100|0.076|
|Average Precision (AP)|0.50:0.95|small|100|0.004|
|Average Precision (AP)|0.50:0.95|medium|100|0.195|
|Average Precision (AP)|0.50:0.95|large|100|0.308|
|Average Recall (AR)|0.50:0.95|all|1|0.082|
|Average Recall (AR)|0.50:0.95|all|10|0.167|
|Average Recall (AR)|0.50:0.95|all|100|0.183|
|Average Recall (AR)|0.50:0.95|small|100|0.017|
|Average Recall (AR)|0.50:0.95|medium|100|0.299|
|Average Recall (AR)|0.50:0.95|large|100|0.394|

However, if I evaluate the saved predictions with pycocotools directly, I get different scores: 
|Metric|IoU|Area|MaxDets|Value|
|------|---|----|-------|-----|
|Average Precision (AP)|0.50:0.95|all|100|0.096|
|Average Precision (AP)|0.50|all|100|0.225|
|Average Precision (AP)|0.75|all|100|0.061|
|Average Precision (AP)|0.50:0.95|small|100|0.004|
|Average Precision (AP)|0.50:0.95|medium|100|0.156|
|Average Precision (AP)|0.50:0.95|large|100|0.246|
|Average Recall (AR)|0.50:0.95|all|1|0.066|
|Average Recall (AR)|0.50:0.95|all|10|0.134|
|Average Recall (AR)|0.50:0.95|all|100|0.147|
|Average Recall (AR)|0.50:0.95|small|100|0.014|
|Average Recall (AR)|0.50:0.95|medium|100|0.239|
|Average Recall (AR)|0.50:0.95|large|100|0.315|

**Reproduction**
Here are my test_evaluator configuration and pycocotools evaluation code snippet.
```python
test_evaluator = dict(
    ann_file='/workspace/datasets/synthia/rand_cityscapes/coco/coco_test_tiny.json',
    backend_args=None,
    format_only=False,
    metric='bbox',
    outfile_prefix='/workspace/output/yolov3_save_test_results/predictions',
    proposal_nums=[100,1,10],
    type='CocoMetric')
```
```python
def evaluate(ann_file, pred_file):
    # Load ground truth
    coco_gt = COCO(ann_file)
    # Load predictions
    coco_dt = coco_gt.loadRes(pred_file)
    coco_eval = COCOeval(coco_gt, coco_dt, iouType='bbox')
    coco_eval.evaluate()
    coco_eval.accumulate()
    coco_eval.summarize()
```

**Environment**
I am developing in a Docker container and using *pycocotools* version 2.0.7. Here is an output of ```python mmdet/utils/collect_env.py```:
```
sys.platform: linux
Python: 3.7.13 (default, Mar 29 2022, 02:18:16) [GCC 7.5.0]
CUDA available: True
MUSA available: False
numpy_random_seed: 2147483648
GPU 0: NVIDIA GeForce GTX 1080 Ti
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 11.3, V11.3.109
GCC: gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
PyTorch: 1.12.1
PyTorch compiling details: PyTorch built with:
  - GCC 9.3
  - C++ Version: 201402
  - Intel(R) oneAPI Math Kernel Library Version 2021.4-Product Build 20210904 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.6.0 (Git Hash 52b5f107dd9cf10910aaa19cb47f3abf9b349815)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.3
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_37,code=compute_37
  - CuDNN 8.3.2  (built against CUDA 11.5)
  - Magma 2.5.2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.3, CUDNN_VERSION=8.3.2, CXX_COMPILER=/opt/rh/devtoolset-9/root/usr/bin/c++, CXX_FLAGS= -fabi-version=11 -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Werror=cast-function-type -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.12.1, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=OFF, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_ROCM=OFF, 

TorchVision: 0.13.1
OpenCV: 4.10.0
MMEngine: 0.10.4
MMDetection: 3.3.0+

**Comment by daniellehot** [comment]
Solved it. I forgot to explicitly define the metainfo key of the dataloader dictionary. Here is my code snippet for generating dataloaders before and after. ```dataset['classes']``` is a list of class names, e.g. ['car', 'bus', 'train'].

BEFORE
```python
def get_dataloader(cfg: Config, train_test_val: str, dataset: dict):
    _options = ['train', 'val', 'test']
    assert train_test_val in _options , f"{train_test_val} must be one of {_options}"

    dataloader_dict = deepcopy(cfg[f"{train_test_val}_dataloader"])
    pipeline = _get_pipeline(train_test_val)

    dataloader_dict['dataset'].update(
        type=dataset['type'],
        ann_file = dataset[f'{train_test_val}_json'],
        data_prefix = dict(img = dataset['data_prefix']),
        data_root = dataset['data_root'],
        pipeline = pipeline
    )

    dataloader_dict.update(
        batch_size = BATCH_SIZE,
        num_workers = NUM_WORKERS
    )

    return dataloader_dict
```

AFTER (I found the fix here https://mmdetection.readthedocs.io/en/latest/advanced_guides/customize_dataset.html)
```python
def get_dataloader(cfg: Config, train_test_val: str, dataset: dict):
    _options = ['train', 'val', 'test']
    assert train_test_val in _options , f"{train_test_val} must be one of {_options}"

    dataloader_dict = deepcopy(cfg[f"{train_test_val}_dataloader"])
    pipeline = _get_pipeline(train_test_val) 

    dataloader_dict['dataset'].update(
        type=dataset['type'],
        metainfo=dict(classes = tuple(dataset['classes'])), # <--- FIX 
        ann_file = dataset[f'{train_test_val}_json'],
        data_prefix = dict(img = dataset['data_prefix']),
        data_root = dataset['data_root'],
        pipeline = pipeline
    )

    dataloader_dict.update(
        batch_size = BATCH_SIZE,
        num_workers = NUM_WORKERS
    )

    return dataloader_dict
``` 

I am going to close the issue now.
